readme.md
# IsoNet

## Introduction
`IsoNet` is an R package for isotopologue similarity networking to discover unknown reactions or metabolites.

The docker image (https://hub.docker.com/r/zhulab/isonet-r) contains entire environment for running `IsoNet`. For convenience and taking fully use of `IsoNet`, users can pull it and run `IsoNet` just as following.

## What is IsoNet-r

`IsoNet-r` is a Docker environment to build isotopologue similarity network with `IsoNet` R package. It is based on the [`r-base`](https://hub.docker.com/_/r-base/) docker.

## Pulling image

Users can pull the IsoNet-r image with the following script

```bash
docker pull zhulab/IsoNet-r
```

## Data preparation

The data folder should contain isotopologue pattern table (.csv), the identification table of labeled metabolite (.csv) and MS2 sprectra (.msp). Demo files could be downloaded from [https://www.zhulab.cn/usr/uploads/misc/isonet/IsoNet_demo_data.zip].


## R script preparation
To run the data processing, an R script named [run.R](extra/run.R) should be placed in the data folder.
Here we provide an example. Users only need to change tracer_table_file,  ms2_file,  id_table, and sample_names . Other parameters are recommended parameters.
  
```R
library(IsoNet)
wd <- '.'
setwd(wd)

constructMIDNet(tracer_table_file = "./Result/isotopologue_pattern_table.csv",
                ms2_file = "ms2_data.msp",
                id_table = "./Result/ID_labeled_metabolites.csv",
                sample_names = c("GGA_1","GGA_2","GGA_3","GGA_4","GGA_5","GGA_6"),
                dir_path = '.',
                mid_cutoff = 0.7,
                mid_fc = 20,
                mid_isoDegree = 0.1,
                mid_min_motifLen = 0.5,
                ms2_score_cutoff = 0.5,
                mid_max_motif = 1000L,
                ignore_max = TRUE,
                scoring_approach = 'gnps',
                mass_diff_freq_cut_off = 4L)
```

**Import parameters**

- `tracer_table_file`: the file name of isotopologue pattern table which contains the labeled fraction of each isotopologue.
- `ms2_file`: the file name of ms2 file.
- `id_table`: the identification table name of labeled metabolites.
- `sample_names`: The sample names in tracer_table_file that be used to calculate isotopologue pattern similarity.
- `mid_cutoff`: the isotopogue pattern similarity score cutoff
- `ms2_score_cutoff`：the MSMS similarity score cutoff
- `mid_fc`: the labeled fraction radio between the highest and the second highest isotopologues to class the type II and type III.
- `mid_isoDegree`: The cutoff for the minimum labeling fraction of highest isotopologue in the motif.
- `mid_min_motifLen`：In Type III, The minimum carbon number of generated motif. For example, setting it to 0.5 indicates that the carbon number of the motif should be at least 50% longer than that of the metabolite with more carbons.
- `mid_max_motif`：In Type III, the maximum number of generated motifs.
- `ignore_max`：In Type I, whether to consider the position of the highest isotopologue. Setting it to TRUE indicates that it is not considered, and the calculation is performed directly.
- `scoring_approach`：The algorithm used for scoring the MSMS simlairty.
- `mass_diff_freq_cut_off`: The threshold for delta masses used for edge annotation based on its frequency in the KEGG database.


## Run data processing work with IsoNet-r image

- go to your data folder (e.g., data)

```base
cd data
```

- run docker using following code (*User should be permitted to run docker service*)

```bash
# MUST keep the code exactly as it is!
docker run -it --rm -v "$PWD":/data -u $(id -u ${USER}):$(id -g ${USER}) zhulab/IsoNet-r Rscript run.R
```

- wait till data processing work done

- Explaining `docker run` arguments
  
  - `-v "$PWD":/home/${USER}`: mapping current directory as home directory in docker container
  
  - `-u $(id -u ${USER}):$(id -g ${USER})`: using current user to run the container
  
  - `Rscript ~/run.R`: run run.R in container home directory with `Rscript`  command

## The result 

The main results are listed following:
  - labeled metabolite pairs named "Labeled metabolite pairs.csv". It can be imported into Cytoscape to built a network.


# License
<a rel="license" href="https://creativecommons.org/licenses/by-nc-nd/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-nd/4.0/88x31.png" /></a>
  
  This work is licensed under the Attribution-NonCommercial-NoDerivatives 4.0 International (CC BY-NC-ND 4.0)


NEWS
## Version 0.1.00 (20220902)
-------------
o initial release version


## Version 0.1.01 (20220906)
-------------
o fix bugs: untransfered type parameter lead unconsistent MID score
o fix bugs: no edge label


## Version 0.1.02 (20220923)
-------------
o fix bugs: filtering MS2 matched pairs using n_frag_total instead of n_frag_match
o add parallel processing function


## Version 0.1.03 (20220924)
-------------
o change parallel running function


## Version 0.1.04 (20221202)
-------------
o change consin to Manhattan MID similarity caculation methods


## Version 0.1.05 (20221223)
-------------
o change the MID similarity methods of features with same C number. 
  type II to type I
  
  
## Version 0.1.06 (20230131)
-------------
o edge annotation tolerance 0.002 Da.
o change library table(KEGG).
o output reaction class and freq.


## Version 0.2.01 (20230131)
-------------
o add ID results to the final table.
o add reaction step (calculateStep)


## Version 0.2.02 (20230203)
-------------
o fix output bug of MID calculation(motif source and type)
o change library table(KEGG).simple name,redefined reaction class,C_diff
o output C_diff.


## Version 0.2.03 (20230209)
-------------
o fix output bug of MID calculation(change same C carbon number judgement code module position,include type I and type II with same C number)
o update delta mass lib (unique mass,more than 1)


## Version 0.2.04 (20230214)
-------------
o consider the adduct forms in function addEdgeLabel
o output all reaction classes
o add parameter to trim the mass difference table (mass_diff_freq_cut_off)
o modify the reaction type


## Version 0.2.05 (20230317)
-------------
o debug ignore_max parameter


## Version 0.2.06 (20230318)
-------------
o Debug mass difference calculation
o Modify output files


## Version 0.2.07 (20230328)
-------------
o Modify the output of the package


## Version 0.2.08 (20230420)
-------------
o Add "|" to the "From" and "To" column
o Modify the path of final result
o Modify the some variates according to the inputted ID table


## Version 0.2.09 (20230420)
-------------
o Debug the polarity recognization during adding edge labels.


## Version 0.2.10 (20230428)
-------------
o Debug mz match during edge annotation.


## Version 0.2.11 (20230924)
-------------
o Add RT filter for edge annotations.


## Version 0.2.12 (20230924)
-------------
o Modify the colnames of the output file.
o Rename the package.


## Version 0.2.13 (20231020)
-------------
o Debug removing features without MS2 during file checking.

tests/test.R
# constructMIDNet(tracer_table_file = 'MID_more_than_3_POS_maxC.csv',
#                 ms2_file = 'ms2_data.msp',
#                 sample_names = c("GGA_1","GGA_2","GGA_3","GGA_4","GGA_5","GGA_6"),
#                 dir_path = "G:/00_projects/06_isotopeTracing/data_processing/20220906_MIDNet",
#                 mid_cutoff = 0.95,
#                 mid_fc = 20,
#                 mid_isoDegree = 0.02,
#                 mid_min_motifLen = "50%",
#                 mid_max_motif = 1000,
#                 scoring_approach = 'gnps',
#                 mz_tol_ms2 = 15,
#                 ms2_score_cutoff = 0.5,
#                 num_matched_frag_cutoff = 1)
#
#
################################################################################
#
# load('G:/00_projects/06_isotopeTracing/data_processing/20220831_MIDNet/00_intermidate_data/list_SIMres.RData')
#
# list_SIMres
# which(NodeTable$From == 'M292T447NA5' & NodeTable$To == 'M380T404C03451')
#
# NodeTable[501567,]
# NodeTable[248382,]

# library(CalMIDSIM)
# setwd("D:/gaoyang/MID_caculation_test/20220719_POS_293T")
#
# calMIDSim(NodeTable = NULL,
#           TargetTable_dir = "G:/00_projects/06_isotopeTracing/test/MID_more_than_3_POS_maxC.csv",
#           TargetSample = c("GGA_1","GGA_2","GGA_3","GGA_4","GGA_5","GGA_6"),
#           Cutoff = 0.95,
#           FoldChange = 20,
#           IsoDegree = 0.02,
#           MinMotifLength = "60%",
#           printall = FALSE,
#           maxMotif = 1000)
#
#
# NodeTable = NULL
# Pattern_name = "Unknown"
# TargetTable_dir = 'G:/00_projects/06_isotopeTracing/data_processing/20220831_MIDNet'
# TargetSample = c("GGA_1","GGA_2","GGA_3","GGA_4","GGA_5","GGA_6")
# Cutoff = 0.9
# IsoDegree = 0.1
# FoldChange = 20
# MinMotifLength = "50%"
# printall = FALSE
# maxMotif = 10000
#
#
#
# MIDSIMCal(TargetTable_dir = "MID_more_than_3_POS_maxC.csv",
#           TargetSample = c("GGA_1","GGA_2","GGA_3","GGA_4","GGA_5","GGA_6"),
#           Cutoff = 0.95,
#           FoldChange = 20,
#           IsoDegree = 0.02,
#           MinMotifLength = "50%",
#           printall = FALSE,
#           maxMotif = 1000)
#
#
# constructMIDNet(tracer_table_file = "MID_more_than_3_POS_maxC.csv",
#                 ms2_file = "ms2_data.msp",
#                 sample_names = c("GGA_1","GGA_2","GGA_3","GGA_4","GGA_5","GGA_6"),
#                 dir_path = 'G:/00_projects/06_isotopeTracing/data_processing/20220831_MIDNet',
#                 mid_cutoff = 0.95,
#                 mid_fc = 20,
#                 mid_isoDegree = 0.02,
#                 mid_min_motifLen = "50%",
#                 mid_max_motif = 1000)
#
#
# test <- c(pairs$From_feature, pairs$To_feature) %>% unique()
# temp_idx <- which(!(test %in% names(ms2_spec)))
# test <- test[temp_idx]
# pairs <- pairs %>%
#   dplyr::filter(!(From_feature %in% test | To_feature %in% test))
#
#
#
################################################################################
#
# constructMIDNet(tracer_table_file = "MID_more_than_3_POS_maxC.csv",
#                 ms2_file = "ms2_data.msp",
#                 sample_names = c("GGA_1","GGA_2","GGA_3","GGA_4","GGA_5","GGA_6"),
#                 dir_path = 'G:/00_projects/06_isotopeTracing/data_processing/20220831_MIDNet',
#                 multi_thread = 6,
#                 mid_cutoff = 0.95,
#                 mid_fc = 20,
#                 mid_isoDegree = 0.02,
#                 mid_min_motifLen = "50%",
#                 mid_max_motif = 1000)
#
# constructMIDNet(tracer_table_file = "MID_more_than_3_POS_maxC.csv",
#                 ms2_file = "ms2_data.msp",
#                 sample_names = c("GGA_1","GGA_2","GGA_3","GGA_4","GGA_5","GGA_6"),
#                 dir_path = 'G:/00_projects/06_isotopeTracing/data_processing/20220923_MIDNet/',
#                 multi_thread = 6,
#                 mid_cutoff = 0.95,
#                 mid_fc = 20,
#                 mid_isoDegree = 0.02,
#                 mid_min_motifLen = "50%",
#                 mid_max_motif = 1000)


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

# q: what does this code mean

devtools::load_all('e:/Work/dev_r/MIDNet/MIDNet')

R/zzz.R
# Copyright © 2022- ZhuMSLab ALL right reserved
#
# @author: Yan-dong Yin
# @contact: yddream@gmail.com
# @project: MIDNet
# @file: zzz.R
# @time: 2023-04-14 22:51


################################################################################
# startup massage --------------------------------------------------------------
.onAttach <- function(libname, pkgname){
  nmark <- 80
  welcome <- paste0("WELCOME to ", getPackageName(), " v", packageVersion(getPackageName()))
  welcome <- .gen_marks(welcome)
  end_mark <- .gen_marks(nmark = nchar(welcome))
  packageStartupMessage(welcome,
                        "\nIf you have any questions, please send email to zhouzw@sioc.ac.cn.",
                        "\nAuthors: Yang Gao, Mingdu Luo, Hongmiao Wang, Zhiwei Zhou and Dr. Zhengjiang Zhu (jiangzhu@sioc.ac.cn).",
                        "\nMaintainer: Yang Gao",
                        "\n",.gen_marks(content = 'NEWS', mark='-', nmark= nmark),
                        "\nVersion 0.2.13 (20231020)",
                        "\n\to Debug removing features without MS2 during file checking.",
                        "\n",
                        end_mark,
                        "\n")
}

.gen_marks <- function(content=NULL, mark='=', nmark= 80) {
  if (!is.null(content)) {
    mark_content <-  paste0(rep(mark, (nmark - nchar(content))/2-1), collapse = '')
    return(paste(mark_content, content, mark_content))
  } else {
    return(paste0(rep(mark, nmark), collapse = ''))
  }
}

R/Temp.R
################################################################################
# setting style ----------------------------------------------------------------
errorStyle <- crayon::red
warnStyle <- crayon::magenta
spliteStyle <- crayon::silver
startStyle <- crayon::blue
stageStyle <- crayon::yellow
normalStyle <- crayon::green


################################################################################
# 20220901 edge rule -----------------------------------------------------------
# mass_diff_table <- readr::read_csv('G:/00_projects/06_isotopeTracing/01_package/data/20220728_edge_rules.csv')
#
# idx <- which(mass_diff_table$AtomicDifference %in% c('none_adduct',
#                                                      'transformation',
#                                                      'unspecified',
#                                                      'unspecified|unspecified|unspecified',
#                                                      'unspecified|unspecified|unspecified|unspecified|unspecified'))
# mass_diff_table <- mass_diff_table[-idx, ]
#
# mass_diff_table <- mass_diff_table %>%
#   dplyr::mutate(GroupDeltaMass_abs = abs(GroupDeltaMass))
#
#
# usethis::use_data(mass_diff_table, overwrite = TRUE)


R/Main.R
################################################################################
#' @title constructMIDNet
#' @author Zhiwei Zhou
#' @param tracer_table_file
#' @param ms2_file
#' @param dir_path
#' @param mid_cutoff
#' @param mid_fc
#' @param mid_isoDegree
#' @param mid_min_motifLen
#' @param min_max_motif
#' @param scoring_approach
#' @param mz_tol_ms2
#' @param mz_tol_label
#' @param mass_diff_freq_cut_off
#' @import crayon
#' @importFrom magrittr '%>%'
#' @export

# c("GGA_1","GGA_2","GGA_3","GGA_4","GGA_5","GGA_6")

constructMIDNet <- function(tracer_table_file = "MID_more_than_3_POS_maxC.csv",
                            ms2_file = "ms2_data.msp",
                            id_table = 'id_table.csv',
                            sample_names,
                            dir_path = '.',
                            multi_thread = 1,

                            # parameters: calculating MID similarity
                            mid_cutoff = 0.95,
                            mid_fc = 20,
                            mid_isoDegree = 0.02,
                            mid_min_motifLen = 0.6,
                            mid_max_motif = 1000,
                            ignore_max = FALSE,

                            # parameters: calculating ms2 network
                            scoring_approach = 'gnps',
                            mz_tol_ms2 = 15,
                            ms2_score_cutoff = 0.5,
                            num_matched_frag_cutoff = 1,

                            # parameters: add edge labels
                            mz_tol_label = 15,
                            mass_diff_freq_cut_off = 1
                            ) {
  cat(startStyle("IsoNet run log\n"))
  cat(startStyle(as.character(Sys.time()), '\n\n'))

  # create run.log file
  cat("IsoNet run log\n", file = file.path(dir_path, "run.log.txt"))
  cat('IsoNet verison:', as.character(packageVersion('IsoNet')), '\n', file = file.path(dir_path, "run.log.txt"), append = TRUE)
  cat(as.character(Sys.time()), '\n', file = file.path(dir_path, "run.log.txt"), append = TRUE)


  # check file existed
  cat(spliteStyle("===============================================\n"))
  cat(stageStyle("Step 1: check required files ...\n"))
  cat("Step 1: check required files ...\n", file = file.path(dir_path, "run.log.txt"), append = TRUE)
  checkFiles(dir_path = dir_path,
             tracer_table_file = tracer_table_file,
             ms2_file = ms2_file,
             id_table = id_table)

  # construct MID network
  cat(spliteStyle("===============================================\n"))
  cat(stageStyle("Step 2: calculate isotopologue similarity ...\n"))
  cat("Step 2: calculate isotopologue similarity ...\n", file = file.path(dir_path, "run.log.txt"), append = TRUE)
  if (!('GlobalMatchResult_filterred.csv' %in% list.files(file.path(dir_path, '00_intermidate_data')))) {
    calMIDSim(NodeTable = NULL,
              TargetTable_dir = file.path('00_intermidate_data', 'trace_table_clean.csv'),
              TargetSample = sample_names,
              dir_path = dir_path,
              multi_thread = multi_thread,
              Cutoff = mid_cutoff,
              FoldChange = mid_fc,
              IsoDegree = mid_isoDegree,
              MinMotifLength = mid_min_motifLen,
              maxMotif = mid_max_motif,
              ignore_max = ignore_max)
  } else {
    cat(warnStyle('Warning: the MID similarity file existed, skip this step!\n'))
  }


  # construct MS2 network
  cat(spliteStyle("===============================================\n"))
  cat(stageStyle("Step 3: calculate molecule network similarity ...\n"))
  cat("Step 3: calculate molecule network similarity ...\n", file = file.path(dir_path, "run.log.txt"), append = TRUE)

  if (!('ms2_sim_table.csv' %in% list.files(file.path(dir_path, '00_intermidate_data')))) {
    calMS2Net(network_file = 'GlobalMatchResult_filterred.csv',
              dir_path = dir_path,
              ms2_file = ms2_file,
              scoring_approach = scoring_approach,
              mz_tol_ms2 = mz_tol_ms2,
              ms2_score_cutoff = ms2_score_cutoff,
              num_matched_frag_cutoff = num_matched_frag_cutoff)
  } else {
    cat(warnStyle('Warning: the MS2 similarity file existed, skip this step!\n'))
  }


  # output
  cat(spliteStyle("===============================================\n"))
  cat(stageStyle("Step 4: export MID networks ...\n"))
  cat("Step 4: export MID networks ...\n", file = file.path(dir_path, "run.log.txt"), append = TRUE)

  # Node table
  addNodeLabel(tracer_table_file = 'trace_table_clean.csv',
               network_file = 'ms2_sim_table_filter.csv',
               dir_path = dir_path)

  # Edge table
  addEdgeLabel(tracer_table_file = 'trace_table_clean.csv',
               network_file = 'ms2_sim_table_filter.csv',
               id_table = id_table,
               dir_path = dir_path,
               mz_ppm = mz_tol_label,
               mass_diff_freq_cut_off = mass_diff_freq_cut_off)

  # Add ID and Step
  calculateStep(id_table = id_table,
                dir_path = dir_path)


  cat("===============================================\n")
  cat(stageStyle("Done!\n"))
  cat(startStyle(as.character(Sys.time()), '\n\n'))
  cat(as.character(Sys.time()), file = file.path(dir_path, "run.log.txt"), append = TRUE)

}

################################################################################
# checkFiles -------------------------------------------------------------------

checkFiles <- function(dir_path = '.',
                       tracer_table_file = "G:/00_projects/06_isotopeTracing/test/MID_more_than_3_POS_maxC.csv",
                       ms2_file = "ms2_data.msp",
                       id_table = id_table
){
  # browser()
  # check existences
  list_file_1 <- list.files(dir_path, recursive = F, full.names = F)
  list_file <- list.files(file.path(dir_path, 'Result'), recursive = F, full.names = T)
  if (!(tracer_table_file %in% list_file)) {
    stop('The tracer table', tracer_table_file, 'NOT existed!\n')
  }
  if (!(ms2_file %in% list_file_1)) {
    stop(errorStyle('The MS2 file', ms2_file, 'NOT existed!\n'))
  }

  if (!(id_table %in% list_file)) {
    stop(errorStyle('The id table ', id_table, 'NOT existed!\n'))
  }

  cat(normalStyle('Required files: Checked!\n'))

  # check format
    # tracer_table
  temp_data <- readr::read_csv(file.path(dir_path, tracer_table_file)
                               # , show_col_types = FALSE
                               )[,-1] # skip the first useless column (20230420)
  temp <- match(c('name', 'id', 'mz', 'formula', 'compound', 'label'), colnames(temp_data))
  if (!identical(temp, 1:6)) {
    stop(errorStyle('Plese check tracer_table_file format\n'))
  }

  if (!is.numeric(temp_data$label)) {
    stop(errorStyle('The label column is not numeric\n'))
  }
  cat(normalStyle('Tracer table format check: Checked!\n'))

  # ms2_file
  if (!(all(stringr::str_detect(ms2_file, pattern = '(\\.msp|\\.mgf)')))) {
    stop(errorStyle('IsoNet only support mgf/msp formats.'))
  }
  cat(normalStyle('MS2 file format check: Checked!\n'))

  # remove features without ms2
  ms2_spec <- readMSP(file.path(dir_path, ms2_file), mode = 'all')
  names(ms2_spec) <- sapply(ms2_spec, function(x){
    x$info$NAME
  })
  idx_rm <- which(!(temp_data$name %in% names(ms2_spec)))
  if (length(idx_rm) == 0) {
    cat(normalStyle('All features have MS/MS spectra: Checked!\n'))

    dir.create(file.path(dir_path, '00_intermidate_data'), showWarnings = FALSE, recursive = TRUE)
    readr::write_csv(temp_data,
                     file = file.path(dir_path, '00_intermidate_data', 'trace_table_clean.csv'))
  } else {
    if (length(idx_rm) == nrow(temp_data)) {
      stop(errorStyle('No features have MS2 spectrum!\n'))
    } else {
      cat(warnStyle('Warning: ', length(idx_rm), 'features without MS2 are removed.'))
    }

    dir.create(file.path(dir_path, '00_intermidate_data'), showWarnings = FALSE, recursive = TRUE)
    # temp_data <- temp_data[-idx,,drop = FALSE](20231020)
    temp_data <- temp_data[-idx_rm,,drop = FALSE]
    readr::write_csv(temp_data,
                     file = file.path(dir_path, '00_intermidate_data', 'trace_table_clean.csv'))
  }

}

################################################################################
# Output -----------------------------------------------------------------------
  # addEdgeAnnotation ----------------------------------------------------------
################################################################################
#' @title addEdgeLabel
#' @import stringr
addEdgeLabel <- function(tracer_table_file,
                         network_file = 'ms2_sim_table.csv',
                         id_table = id_table,
                         dir_path = '.',
                         mz_ppm = 15,
                         mass_diff_freq_cut_off = 1){

  # browser()

  tracer_table <- readr::read_csv(file.path(dir_path, '00_intermidate_data', tracer_table_file)
                                  # , show_col_types = FALSE
                                  )
  network_table <- readr::read_csv(file.path(dir_path, '00_intermidate_data', network_file)
                                   # , show_col_types = FALSE
                                   )
  id_table <- readr::read_csv(file.path(dir_path, id_table)
                              # , show_col_types = FALSE
  )

  mass_diff_table <- mass_diff_table[which(mass_diff_table$Freq >= mass_diff_freq_cut_off), ]

  # match the network_table table and id_table table to get the mz
  idx_from <- match(network_table$From_feature, id_table$name)
  network_table$from_mz <- id_table$mz[idx_from]
  idx_to <- match(network_table$To_feature, id_table$name)
  network_table$to_mz <- id_table$mz[idx_to]
  # network_table$mass_diff_abs <- abs(network_table$to_mz - network_table$from_mz)
  # network_table$mass_diff_abs_upper <- network_table$mass_diff_abs * (1 + mz_ppm * 10 ^ (-6))
  # network_table$mass_diff_abs_lower <- network_table$mass_diff_abs * (1 - mz_ppm * 10 ^ (-6))

  # the mass difference should consider the adduct forms
  idx <- str_detect(string = id_table$confidence_level, pattern = 'level1|level2') # get the level 1 and level 2 id 
  id_table_level12 <- id_table[which(idx), ]
  rm(idx)

  # idx_from <- match(network_table$From_feature, id_table_level12$peak_name)
  # idx_to <- match(network_table$To_feature, id_table_level12$peak_name)

  # the feature name of id table had been changed (20230420)
  idx_from <- match(network_table$From_feature, id_table_level12$name)
  idx_to <- match(network_table$To_feature, id_table_level12$name)
  
  network_table$temp_adduct_from <- id_table_level12$adduct[idx_from]
  network_table$temp_adduct_to <- id_table_level12$adduct[idx_to]

  if(any(stringr::str_detect(id_table_level12$adduct, '\\[M\\+H\\]\\+'))){
    network_table$temp_adduct_from[which(is.na(network_table$temp_adduct_from))] <- '[M+H]+'
    network_table$temp_adduct_to[which(is.na(network_table$temp_adduct_to))] <- '[M+H]+'
    pos <- 'positive'
  }else{
    network_table$temp_adduct_from[which(is.na(network_table$temp_adduct_from))] <- '[M-H]-'
    network_table$temp_adduct_to[which(is.na(network_table$temp_adduct_to))] <- '[M-H]-'
    pos <- 'negative'
  }


  network_table$temp_adduct_from <- sapply(seq(nrow(network_table)), function(i){
    return(strsplit(network_table$temp_adduct_from[i], ';')[[1]][1])
  })

  network_table$temp_adduct_to <- sapply(seq(nrow(network_table)), function(i){
    return(strsplit(network_table$temp_adduct_to[i], ';')[[1]][1])
  })


  if(pos == 'positive'){
    network_table$temp_adduct_from[which(network_table$temp_adduct_from == '[M]+')] <- '[M+H]+'
    network_table$temp_adduct_to[which(network_table$temp_adduct_to == '[M]+')] <- '[M+H]+'
  }else{
    network_table$temp_adduct_from[which(network_table$temp_adduct_from == '[M-2H]-')] <- '[M-H]-'
    network_table$temp_adduct_to[which(network_table$temp_adduct_to == '[M-2H]-')] <- '[M-H]-'
  }

  lib_adduct_nl <- lib_adduct_nl[[pos]]
  network_table$delta_from <- lib_adduct_nl$delta_mz[match(network_table$temp_adduct_from, lib_adduct_nl$adduct)]
  network_table$delta_to <- lib_adduct_nl$delta_mz[match(network_table$temp_adduct_to, lib_adduct_nl$adduct)]


  network_table$temp_exact_mass_from <- sapply(seq(nrow(network_table)), function(i){
    return(calculateMz(mz = network_table$from_mz[i],
                       adduct = network_table$temp_adduct_from[i],
                       delta_mz = network_table$delta_from[i]))
  }, simplify = TRUE)

  network_table$temp_exact_mass_to <- sapply(seq(nrow(network_table)), function(i){
    return(calculateMz(mz = network_table$to_mz[i],
                       adduct = network_table$temp_adduct_to[i],
                       delta_mz = network_table$delta_to[i]))
  }, simplify = TRUE)


  network_table$mass_diff_abs <- abs(as.numeric(network_table$temp_exact_mass_from) - as.numeric(network_table$temp_exact_mass_to))

  network_table$temp_adduct_from <- NULL
  network_table$temp_adduct_to <- NULL
  network_table$delta_from <- NULL
  network_table$delta_to <- NULL
  network_table$temp_exact_mass_from <- NULL
  network_table$temp_exact_mass_to <- NULL

  network_table$mass_diff_abs_upper <- as.numeric(network_table$mass_diff_abs + 0.002)
  network_table$mass_diff_abs_lower <- as.numeric(network_table$mass_diff_abs - 0.002)
  res_atom_diff <- lapply(seq(nrow(network_table)), function(i){
    temp_row <- network_table[i, ]
    temp_idx <- which(mass_diff_table$GroupDeltaMass_abs >= temp_row$mass_diff_abs_lower &
                        mass_diff_table$GroupDeltaMass_abs <= temp_row$mass_diff_abs_upper)
    # mass_accuracy <- round(abs(mass_diff_table$GroupDeltaMass_abs[temp_idx] - temp_row$mass_diff_abs) / temp_row$mass_diff_abs * 10 ^6,
    #                        digits = 0)
    DetlaMassAccuracy <- round(abs(mass_diff_table$GroupDeltaMass_abs[temp_idx] - temp_row$mass_diff_abs), digits = 4)

    # res_mass_diff_thero <- paste0(mass_diff_table$GroupDeltaMass[temp_idx],
    #                               collapse = ';')

    DetlaMassAccuracy <- paste0(DetlaMassAccuracy,
                                collapse = ';')

    AtomDiff <- paste0(mass_diff_table$AtomDifference[temp_idx],
                                 collapse = ';')

    ReactionFreq <- paste0(mass_diff_table$Freq[temp_idx],
                                 collapse = ';')

    ReactionClass <- paste0(mass_diff_table$reaction_class[temp_idx],
                                 collapse = ';')

    ReactionClassAll <- paste0(mass_diff_table$reaction_class_all[temp_idx],
                             collapse = ';')

    # with_C_diff <- paste0(mass_diff_table$with_C_diff[temp_idx],
    #                          collapse = ';')

    return(data.frame(DetlaMassAccuracy = DetlaMassAccuracy,
                      AtomDiff = AtomDiff,
                      ReactionClass = ReactionClass,
                      ReactionClassAll = ReactionClassAll,
                      ReactionFreq = ReactionFreq,
                      stringsAsFactors = FALSE))
  })

  res_atom_diff <- res_atom_diff %>% dplyr::bind_rows()
  network_table$mass_diff_abs_upper <- NULL
  network_table$mass_diff_abs_lower <- NULL
  colnames(network_table)[which(colnames(network_table) == 'mass_diff_abs')] <- 'DetlaMassAbs'

  # result <- network_table %>%
  #   dplyr::bind_cols(res_atom_diff) %>%
  #   dplyr::mutate(edge_name = paste0(From_feature," (interacts with) ", To_feature))

  result <- network_table %>%
    dplyr::bind_cols(res_atom_diff)

  # dir.create(file.path(dir_path, ), showWarnings = FALSE, recursive = TRUE)
  # readr::write_csv(result,
  #                  file = file.path(dir_path, 'edge_table_with_label.csv'))
  readr::write_csv(result,
                   file = file.path(dir_path, '00_intermidate_data', 'edge_table_with_label.csv'))

}


  # addNodeLabel ---------------------------------------------------------------

addNodeLabel <- function(tracer_table_file,
                         network_file = 'ms2_sim_table.csv',
                         dir_path = '.'){
  tracer_table <- readr::read_csv(file.path(dir_path,  '00_intermidate_data', tracer_table_file)
                                  # , show_col_types = FALSE
                                  )
  network_table <- readr::read_csv(file.path(dir_path, '00_intermidate_data', network_file)
                                   # , show_col_types = FALSE
                                   )
  node_list <- c(network_table$From_feature, network_table$To_feature) %>% unique()
  node_table <- match(node_list,
                      tracer_table$name) %>%
    tracer_table[.,] %>%
    dplyr::select(name:label)

  idx <- node_table$id %>%
    stringr::str_detect(pattern = 'NA') %>%
    which()

  node_table$attribute <- 'seed'
  node_table$attribute[idx] <- 'unknown'

  # dir.create(file.path(dir_path), showWarnings = FALSE, recursive = TRUE)
  # readr::write_csv(node_table,
  #                  file = file.path(dir_path, 'node_table_with_label.csv'))

}

################################################################################
#' @title calculateMz
#' @import stringr
calculateMz <- function(mz,
                        adduct,
                        delta_mz,
                        nmol = NULL,
                        ncharge = NULL){
  if (length(nmol) == 0) {
    if (stringr::str_detect(adduct, pattern = '2M')) {
      exact_mass <- (mz - delta_mz)/2
    } else if (stringr::str_detect(adduct, pattern = '3M')) {
      exact_mass <- (mz - delta_mz)/3
    } else {
      exact_mass <- mz - delta_mz
    }
  } else {
    exact_mass <- (mz - delta_mz)/nmol
  }

  exact_mass
}

